### PR TITLE
Feature/#31 data signal visualization from the node server v1.0

### PR DESCRIPTION
**Pull request by leon-juenemann**
plot_signals

### CODE REVIEW

**Comment by leon-juenemann** [comment]
I build upon this a bit and created a new branch. It changes quite a bit and I didn't want to crash this PR ;)

https://github.com/amosproj/amos2022ws03-software-oscilloscope/pull/58

**Review by nicolaskolbenschlag — APPROVED** [review]
With this implementation the signals for each channel are fixed and can not be parameterized via environment variables anymore. But as this was no requirement, it is fine for me.

File: Apps/backend/src/main.js [hunk 1]
```diff
@@ -1,25 +1,31 @@
 import dgram from "node:dgram";
+import WebSocket, {WebSocketServer} from 'ws';
 
 const server = dgram.createSocket("udp4");
+const socket = new WebSocketServer({
+  port: 9000,
```
Comment (leandertolksdorf) [inline Apps/backend/src/main.js:6]: Maybe it's better to have the WebSocket use ENV vars for host and port like the UDP socket.

File: Apps/backend/src/main.js [hunk 2]
```diff
@@ -1,25 +1,31 @@
...
+  client.send("connection established")
+});
 
 server.on("error", (err) => {
   console.log(`Server error:\n${err.stack}`);
-  server.close();
+  server.close(package_recieved);
 });
 
 server.on("message", (msg, rinfo) => {
```
Comment (leandertolksdorf) [inline Apps/backend/src/main.js:23]: Unused var rinfo

File: Apps/frontend/src/views/Oscilloscope.svelte [hunk 3]
```diff
@@ -1,45 +1,39 @@
...
-    const canvasHeight = 500;
-    const numIntervalsX = 20;
-    const numIntervalsY = 10;
+  let waveElement;
+  let scaleY = 1; // 1V per horizontal line
+  let socket;
 
-    let waveElement;
+  onMount(() => {
+    socket = new WebSocket("ws://localhost:9000");
```
Comment (leandertolksdorf) [inline Apps/frontend/src/views/Oscilloscope.svelte:12]: When using ws://backend:9000, it uses docker's internal network and you don't need to expose port 9000 on the backend.

File: Apps/frontend/src/views/SineWave.svelte [hunk 4]
```diff
@@ -1,35 +1,58 @@
...
+  let channel_samples = Array.from(Array(numberOfChannels), () => new Array(CANVAS_WIDTH).fill(0.0))
+  let lines = []
+
+  let x = 0
+  export const updateBuffer = (samples) => {
+    for (let channelIndex = 0; channelIndex < channel_samples.length; channelIndex++)
+    {
+      channel_samples[channelIndex][x] = samples[channelIndex]
+    }
+    x++;
```
Comment (leandertolksdorf) [inline Apps/frontend/src/views/SineWave.svelte:20]: In this case x = (x+1) % CANVAS_WIDTH would be more readable imo.

**Review by leandertolksdorf — CHANGES_REQUESTED** [review]
Some minor remarks at the JS part, cannot contribute much in Rust.

**Comment by leandertolksdorf** [comment]
The plotting looks weird on my machine...

https://user-images.githubusercontent.com/55246283/202125467-953c2607-684e-4956-88e9-5756a90dd663.mov

**Pull request merged**